src/Analyser/TypeSpecifier.php

```diff
@@ -1419,15 +1419,19 @@ private function processBooleanConditionalTypes(Scope $scope, SpecifiedTypes $le
 	private function findTypeExpressionsFromBinaryOperation(Scope $scope, Node\Expr\BinaryOp $binaryOperation): ?array
 	{
 		$leftType = $scope->getType($binaryOperation->left);
-		$rightType = $scope->getType($binaryOperation->right);
 		if (
 			$leftType instanceof ConstantScalarType
+			&& !$binaryOperation->right instanceof Node\Scalar\String_
```

Comment:
Eh, what's so special about string? Why did you add this change?

Reply:
its required to make https://github.com/staabm/phpstan-src/blob/1.10.x/tests/PHPStan/Analyser/data/bug-6901.php pass

otherwise - because of switch beeing re-written to equal-comparisons and the new return type extension - `$leftType` and `$rightType`  will resolve to constant strings and the method - as it was before the PR - would not specify the right expression

Reply:
That feels like a wrong placement for such fix.

`gettype` is handled in TypeSpecifier around line 1072. Wouldn't it help to first check for `gettype` condition and only after that to perform the thing on line 1424?

Reply:
I think we will have the same problem for other functions beeing used in a switch statement/identical comparison and have a return type extension and a type specifying extension with constant scalars.

we cannot detect it in line 1072, as the problem is not within the `FuncCall` itself, but the comparison.

see the bug in action in https://phpstan.org/r/d00e29f6-3163-414b-b83d-f591fd1eaf7d
(I have added a separate test-case for it)

Reply:
If you prefer we could separate that bugfix in a new PR

Reply:
Yeah https://phpstan.org/r/d00e29f6-3163-414b-b83d-f591fd1eaf7d shouldn't be happening, but I'm not sure if it's the same bug...

Reply:
ok, turned it into a separate bug: https://github.com/phpstan/phpstan/issues/9404

Reply:
I had another look at it.

you were right that https://github.com/phpstan/phpstan/issues/9404 does not related to the problem here.

I think the fix is right though and the description in https://github.com/phpstan/phpstan-src/pull/2437/files#r1220239214 is accurate

Reply:
to simplify the test in question:


```
<?php

use function PHPStan\Testing\assertType;

/**
 * @param int $y
 */
function foo($y): void
{
	switch (gettype($y)) {
		case "integer":
			assertType('int', $y);
			break;
		default:
			assertType('*NEVER*', $y);
	}
}
```

runs into 
```
------ ------------------------------------ 
  Line   tst.php                             
 ------ ------------------------------------ 
  15     Expected type *NEVER*, actual: int  
 ------ ------------------------------------ 
```

in this PR, when we drop the `... instanceof Node\Scalar\String_` changes